src/psyclone/domain/lfric/lfric_invoke.py

```diff
@@ -0,0 +1,331 @@
+        # and inter-grid operations
+        self.meshes = DynMeshes(self, self.psy_unique_vars)
+
+        # Initialise the object holding information on any boundary-condition
+        # kernel calls
+        self.boundary_conditions = DynBoundaryConditions(self)
+
+        # Information on all proxies required by this Invoke
+        self.proxies = DynProxies(self)
+
+        # Run-time checks for this invoke
+        self.run_time_checks = LFRicRunTimeChecks(self)
+
+        # Information required by kernels that operate on cell-columns
+        self.cell_iterators = DynCellIterators(self)
+
+        # Information on the required properties of the reference element
+        self.reference_element_properties = DynReferenceElement(self)
+
+        # Properties of the mesh
+        self.mesh_properties = LFRicMeshProperties(self)
+
+        # Manage the variables used to store the upper and lower limits of
+        # all loops in this Invoke.
+        self.loop_bounds = LFRicLoopBounds(self)
+
+        # Extend arg list with stencil information
+        self._alg_unique_args.extend(self.stencil.unique_alg_vars)
+
+        # adding in qr arguments
+        self._alg_unique_qr_args = []
+        for call in self.schedule.kernels():
+            for rule in call.qr_rules.values():
+                if rule.alg_name not in self._alg_unique_qr_args:
+                    self._alg_unique_qr_args.append(rule.alg_name)
+        self._alg_unique_args.extend(self._alg_unique_qr_args)
+        # we also need to work out the names to use for the qr
+        # arguments within the psy layer. These are stored in the
+        # _psy_unique_qr_vars list
+        self._psy_unique_qr_vars = []
+        for call in self.schedule.kernels():
+            for rule in call.qr_rules.values():
+                if rule.psy_name not in self._psy_unique_qr_vars:
+                    self._psy_unique_qr_vars.append(rule.psy_name)
+
+        # lastly, add in halo exchange calls and global sums if
+        # required. We only need to add halo exchange calls for fields
+        # since operators are assembled in place and scalars don't
+        # have halos. We only need to add global sum calls for scalars
+        # which have a gh_sum access.
+        if Config.get().distributed_memory:
+            # halo exchange calls
+            const = LFRicConstants()
+            for loop in self.schedule.loops():
+                loop.create_halo_exchanges()
+            # global sum calls
+            for loop in self.schedule.loops():
+                for scalar in loop.args_filter(
+                        arg_types=const.VALID_SCALAR_NAMES,
+                        arg_accesses=AccessType.get_valid_reduction_modes(),
+                        unique=True):
+                    global_sum = DynGlobalSum(scalar, parent=loop.parent)
+                    loop.parent.children.insert(loop.position+1, global_sum)
+
+    def arg_for_funcspace(self, fspace):
```

Comment:
Not your fault, but this function lacks documentation for input argument `fspace`, the `Generation Error` that it raises and the return value. I would kindly request the function be documented. A nice example of a fully documented class is [LFRicArgDescriptor](https://github.com/stfc/PSyclone/blob/master/src/psyclone/domain/lfric/lfric_arg_descriptor.py#L61). In general, the docstring follows this outline:
```
def method(self, input_arg1, input_arg2):
    '''
    This method does  .......

    :param input_arg1: description of this input argument.
    :type  input_arg1: type/class of input_arg1
    :param input_arg2: description of this input argument.
    :type  input_arg2: type/class of input_arg2

    :returns: description of the return argument.
    :rtype: type/class of the return argument

    :raises: <XXXXError>: if the error conditions are fulfilled...
    :raises: <YYYYError>: if the error conditions are fulfilled...

    '''
```
So, for this method the docstring would be something like
```
def arg_for_funcspace(self, fspace):
    '''
    Returns an argument object which is on the requested
    function space. Searches through all Kernel calls in this
    invoke. Currently the first argument object that is found is used.

    :param fspace: the requested function space a kernel argument is on.
    :type fspace: `<psyclone.domain.lfric.FunctionSpace>`

    :returns: the first kernel argument that is on the named function \
                  space and the associated FunctionSpace object.
    :rtype: (:py:class:`psyclone.dynamo0p3.DynKernelArgument`,
             :py:class:`psyclone.domain.lfric.FunctionSpace`)

    :raises GenerationError: if no argument on the requested function space exists.

    '''

```

The easiest way to find the description and type of arguments is to grep for similar functions, or for a function that the method calls and look for docstrings there. This is how I found the description and type of the return argument here - I grepped for [get_arg_on_space](https://github.com/stfc/PSyclone/blob/master/src/psyclone/dynamo0p3.py#L9071).

If that doesn't work, then it is best to find a test that implements this method, break it and add a temporary `print` statement for the type of the function argument. I will use [test_dyninvoke_arg_for_fs](https://github.com/stfc/PSyclone/blob/master/src/psyclone/tests/dynamo0p3_test.py#L686) as an example here:
* Break the test by modifying the `assert` value, e.g. "No argumentZZZZ found on 'wtheta' space";
* Add a `print` statement to return the type of the input argument,  `print(type(fspace))`,  in this method, just below the method header and docstring;
* Run only the modified test, `pytest dynamo0p3_test.py -k test_dyninvoke_arg_for_fs`;
* Read the type, update the docstring, restore the test and remove the `print` statement.